internal/task/task.go

```diff
@@ -0,0 +1,314 @@
+		fileByte, err := os.ReadFile(taskStatsFile)
+		if err != nil {
+			log.Error(err)
+		}
+		err = json.Unmarshal(fileByte, &TaskStatsMap)
+		if err != nil {
+			log.Error(err)
+		}
+	}
+}
+
+// SaveTaskStatsPeriodic runs periodically to save task stats to file system (checkpointing mechanism).
+func SaveTaskStatsPeriodic(ctx context.Context) {
+	// 30 seconds periodic
+	ticker := time.NewTicker(30 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			taskStatsMutex.Lock()
+			tBytes, err := json.Marshal(TaskStatsMap)
+			if err != nil {
+				log.Errorf("error[%v] in marshalling TaskStatsMap for checkpointing", err)
+				log.Error(err)
+			}
+			err = os.WriteFile(taskStatsFile, tBytes, 0644)
+			if err != nil {
+				log.Errorf("error[%v] in writing taskStatsFile file", err)
+			}
+			taskStatsMutex.Unlock()
+		case <-ctx.Done():
+			fmt.Println("stopping periodic save task stats routine")
+			return
+		}
+	}
+}
+
+// getTaskStats returns task stats for particular task.
+func getTaskStats(id string) (TaskStats, error) {
+	taskStatsMutex.Lock()
+	defer taskStatsMutex.Unlock()
+	if v, ok := TaskStatsMap[id]; ok {
+		return v, nil
+	}
+	return TaskStats{}, errors.New("task not found")
+}
+
+// saveTaskStats saves task stats for a particular task.
+func saveTaskStats(id string, ts TaskStats) {
+	taskStatsMutex.Lock()
+	defer taskStatsMutex.Unlock()
+	TaskStatsMap[id] = ts
+}
+
+// Newtask returns new task instance.
+func Newtask() *Task {
+	task := new(Task)
+	// keeping previous time to -N minutes before scheduling interval.
+	task.PreviousRunTime = time.Now().Add(-task.SchedulingInterval)
+	// keeping nextruntime as current time.
+	task.NextRunTime = time.Now()
+	return task
+}
+
+// AddTaskParams methods adds tasks params.
+func (t *Task) AddTaskParams(tp TaskParams) {
+	t.TaskParams = tp
+	t.ID = tp.Config.Name + "$" + tp.Config.RepositoryOwner + "$" + tp.Config.RepositoryName
+}
+
+// AddInterval methods adds tasks scheduling interval.
+func (t *Task) AddInterval(interval time.Duration) {
+	t.SchedulingInterval = interval
+}
+
+// scheduleNextRunOfTask schedules the next run of the task.
+func (task *Task) ScheduleNextRunOfTask() {
+	task.NextRunTime = time.Now().Add(task.SchedulingInterval)
+}
+
+// Start methods starts teh execution of a particular task.
+func (t *Task) Start() error {
+	var ts TaskStats
+	ts, err := getTaskStats(t.ID)
+	// if error reading previous stats , putting default values for task stats map
+	if err != nil {
+		ts.LastIssueTime = t.PreviousRunTime
+		for _, br := range t.Config.Branches {
+			ts.LastCommitTime[br] = t.PreviousRunTime
+		}
+		ts.LastPullRequestNo = 0
+		ts.TaskID = t.ID
+		// saving default task stats for a task.
+		saveTaskStats(t.ID, ts)
+	}
+	gp := gitprovider.NewGitProvider(t.Config.RepositoryHost, t.Config.RepositoryOwner, t.Config.RepositoryName, t.Config.Username, t.DecodeAccessKey)
+
+	// executing for each target in task
+	for _, tar := range t.Targets {
+		// getting new publisher
```

Comment:
Use concurrent goroutines.

Reply:
ok will be adding concurrent go routines